redis_state_sync.go

```diff
@@ -8,21 +10,35 @@ import (
 )
 
 type RedisStateSync struct {
-	heads    map[string]string
-	c        chan *KinesisRecord
-	recs     chan<- *KinesisRecord
-	mut      sync.Mutex
-	pool     *redis.Pool
-	redisKey string
-	ticker   <-chan time.Time
-	wg       sync.WaitGroup
-	modified bool
+	heads       map[string]string
+	acquired    map[string]Unit
+	c           chan *KinesisRecord
+	recs        chan<- *KinesisRecord
+	mut         sync.Mutex
+	pool        *redis.Pool
+	redisPrefix string
+	savePeriod  time.Duration
+	alivePeriod time.Duration
+	wg          sync.WaitGroup
+	modified    bool
+	lock        string
 }
 
 type RedisStateSyncOptions struct {
-	ShardStateSyncOptions
-	RedisURL string
-	RedisKey string
+	SavePeriod  time.Duration
+	AlivePeriod time.Duration
+	RedisURL    string
+	RedisPrefix string
+}
+
+var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+
+func randString(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letters[rand.Intn(len(letters))]
+	}
+	return string(b)
```

Comment:
The main issue is it's also easy to get it wrong when we write our own. For example, looks like we have to call `rand.Seed` otherwise it defaults to `Seed(1)` which means both processes are going to get the same values.

> Random numbers are generated by a Source. Top-level functions, such as Float64 and Int, use a default shared Source that produces a deterministic sequence of values each time a program is run. Use the Seed function to initialize the default Source if different behavior is required for each run. The default Source is safe for concurrent use by multiple goroutines.

http://golang.org/pkg/math/rand/